src/workflow/snakemake/snakemake.py

```diff
@@ -0,0 +1,190 @@
+
+            # Create node rules
+            f.write("nodes = benchmark.get_nodes()\n")
+            f.write("for node in nodes:\n")
+            f.write("    create_node_rule(node, benchmark)\n\n")
+
+        return snakefile_path
+
+    def run_node_workflow(self, node: BenchmarkNode, cores: int = 1, dryrun: bool = False, work_dir: str = os.getcwd(),
+                          **snakemake_kwargs):
+        """
+        Serializes & runs benchmark node workflow using snakemake.
+
+        Args:
+            node (Benchmark): benchmark node to run
+            cores (int): number of cores to run
+            dryrun (bool): validate the snakemake workflow with the benchmark without actual execution
+            work_dir (str): working directory
+            **snakemake_kwargs: keyword arguments to pass to the snakemake engine
+
+        Returns:
+        - Status code of the workflow run.
+        """
+
+        os.makedirs(work_dir, exist_ok=True)
+
+        # Serialize Snakefile for node workflow
+        snakefile = self.serialize_node_workflow(node, work_dir)
+
+        # Prepare the argv list
+        argv = self._prepare_argv(snakefile, cores, dryrun, work_dir, **snakemake_kwargs)
+
+        # Execute snakemake script
+        success = snakemake_cli(argv)
+
+        return success
+
+    def serialize_node_workflow(self, node: BenchmarkNode, output_dir: str = os.getcwd()):
+        """
+        Serializes a Snakefile for a benchmark node.
+
+        Args:
+            node (BenchmarkNode): benchmark node to serialize
+            output_dir (str): output directory for the Snakefile
+
+        Returns:
+        - Snakefile path.
+        """
+        os.makedirs(output_dir, exist_ok=True)
+
+        benchmark_file = node.get_definition_file()
+
+        # Serialize Snakemake file
+        snakefile_path = os.path.join(output_dir, 'Snakefile')
+        with open(snakefile_path, 'w') as f:
+            self._write_snakefile_header(f)
+            self._write_includes(f, INCLUDES)
+
+            # Load benchmark from yaml file
+            f.write(f'node = load_node("{benchmark_file}", "{node.get_id()}")\n\n')
+
+            # Create capture all rule
+            f.write("input_paths = node.get_input_paths()\n")
+            f.write("output_paths = node.get_output_paths()\n")
+            f.write("all_paths = input_paths + output_paths\n\n")
+            f.write("create_all_rule(all_paths)\n\n")
+
+            # Create node rules
+            f.write("create_node_rule(node)\n\n")
+
+        return snakefile_path
+
+    @staticmethod
+    def _write_snakefile_header(f: TextIO):
+        """Write header for the generated Snakefile"""
+
+        f.write("#!/usr/bin/env snakemake -s\n")
+        f.write("##\n")
+        f.write("## Snakefile to orchestrate YAML-defined omnibenchmarks\n")
+        f.write("##\n")
+        f.write(f"## This Snakefile has been automatically generated on {datetime.now()}\n")
```

Comment:
(low prio) Worth incorporating the benchmarker name/email, and benchmark name and version? As parsed from the benchmarking yaml?

Reply:
Great suggestion, I included your suggested information, should be easy to change / adapt in the future.
For now it will look like this:

```
#!/usr/bin/env snakemake -s
#############################################
# Snakefile for Orchestrating YAML-defined OmniBenchmarks
#############################################
# 
# This Snakefile has been automatically generated on 2024-07-22 16:42:56.973076
# 
# Benchmark Details:
# - Name: OmniBenchmark Benchmark_001
# - Version: 1.0
# 
# Author Information:
# - Contact: John Doe at Robinsons lab, john.doe@uzh.ch
#
#############################################
```